### PR TITLE
Search: Simply SingleFieldsVisitor

### PR DESCRIPTION
**Pull request by nik9000**
`SingleFieldsVisitor` is meant to load a single stored field but it
manages to be quite complex to reason about because it inherits from our
"basic" `FieldsVisitor` which is designed to load many fields. This
breaks that inheritance and adds logic to `SingleFieldsVisitor` so it can
be properly stand alone. While this amounts to more lines of code they
ought to be significantly easier to reason about.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/lookup/LeafFieldsLookup.java [hunk 1]
```diff
@@ -142,16 +142,23 @@ private FieldLookup loadFieldData(String name) {
...
-            fieldVisitor.reset(fieldName);
-            try {
-                reader.document(docId, fieldVisitor);
-                fieldVisitor.postProcess(mapperService);
-                List<Object> storedFields = fieldVisitor.fields().get(data.fieldType().name());
-                data.fields(singletonMap(fieldName, storedFields));
-            } catch (IOException e) {
-                throw new ElasticsearchParseException("failed to load field [{}]", e, name);
+            List<Object> values;
+            if (TypeFieldMapper.NAME.equals(data.fieldType().name())) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/lookup/LeafFieldsLookup.java]: This approach will not work in 6.x which has to be compatible with indices created in 5.x but I think it is quite sensible in master.

File: server/src/test/java/org/elasticsearch/index/shard/RefreshListenersTests.java [hunk 2]
```diff
@@ -324,9 +324,9 @@ public void testLotsOfThreads() throws Exception {
                         try (Engine.GetResult getResult = engine.get(get, engine::acquireSearcher)) {
                             assertTrue("document not found", getResult.exists());
                             assertEquals(iteration, getResult.version());
-                            SingleFieldsVisitor visitor = new SingleFieldsVisitor("test");
+                            DocumentStoredFieldVisitor visitor = new DocumentStoredFieldVisitor();
                             getResult.docIdAndVersion().reader.document(getResult.docIdAndVersion().docId, visitor);
-                            assertEquals(Arrays.asList(testFieldValue), visitor.fields().get("test"));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/index/shard/RefreshListenersTests.java]: I suspect I could just call `document(getResult.docIdAndVersion())` here and save some ceremony.

**Review by jimczi — APPROVED** [review]
LGTM

**Review by jpountz — APPROVED** [review]
:+1: Agreed that it is much easier to reason about now.

**Comment by nik9000** [comment]
Thanks for reviewing @jpountz and @jimczi! I'm going to backport this fairly carefully and open a new PR for it because supporting 5.x indices requires a little more care.

**Comment by nik9000** [comment]
Now that I'm looking at the backport more closely, I'm tempted to not pull this back to 6.x at all. `uid` support isn't really compatible with this change and 5.x support looks tricky as well. So, at least for now, I'm going to leave this as a master only change.

**Pull request merged**